### PR TITLE
Add the bootloader tests.

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

**Comment by nomis52** [comment]
This means I can finally withdraw #131

**Comment by nomis52** [comment]
Woot! Up to 75% test coverage.

File: .travis.yml [hunk 1]
```diff
@@ -78,7 +78,8 @@ before_install:
 
 after_success:
  - if [ "$TASK" = "coverity" ]; then cat /home/travis/build/${TRAVIS_REPO_SLUG}/tests/cov-int/scm_log.txt; fi
- - if [ "$TASK" = "coverage" ]; then cd tests && coveralls --gcov /usr/bin/gcov-5 -b . -e include -e harmony -e mocks -e config.h -e system_config -E '.*Test\.cpp$' --gcov-options '\-lp' --follow-symlinks > /dev/null; fi
+ - if [ "$TASK" = "coverage" ]; then cd tests && coveralls --gcov
+   /usr/bin/gcov-5 -b . -e include -e harmony -e mocks -e config.h -e system_config -e test --gcov-options '\-lp' --follow-symlinks > /dev/null; fi
```
Comment (peternewman) [inline .travis.yml]: /usr/bin/gcov-5 is the argument to the --gcov flag, so can we leave it attached to that spacing wise please.

File: tests/tests/BootloaderTestHelper.cpp [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+using testing::Mock;
+using testing::Return;
+using testing::SaveArg;
+using testing::WithArgs;
+using testing::_;
+
+::std::ostream& operator<<(::std::ostream& os, const DFUState& state) {
+  switch (state) {
+    case APP_STATE_IDLE:
+      return os << "APP_STATE_DETACH(0)";
```
Comment (peternewman) [inline tests/tests/BootloaderTestHelper.cpp]: IDLE?

**Comment by peternewman** [comment]
Generally LGTM

File: .travis.yml (hunk 1 above)
Comment (nomis52) [inline .travis.yml]: oh I hadn't noticed this. I've put it all back on one line.

File: tests/tests/BootloaderTestHelper.cpp (hunk 2 above)
Comment (nomis52) [inline tests/tests/BootloaderTestHelper.cpp]: Yup, fixed.

**Comment by peternewman** [comment]
LGTM

**Pull request merged**